modules/relevantDigitalBidAdapter.js

```diff
@@ -0,0 +1,194 @@
+  };
+  if (cfg.complete) {
+    return cfg; // Most common case, we already have the settings we need (and we won't re-read them)
+  }
+  configByBidder[bidder] = cfg;
+  const bidderConfiguration = config.getConfig(bidder) || {};
+
+  // Read settings set by setConfig({ [bidder]: { ... }}) and if not available - from bid params
+  FIELDS.forEach(({ name, checkParams }) => {
+    cfg[name] = bidderConfiguration[name] || cfg[name];
+    if (!cfg[name] && checkParams) {
+      bids.forEach((bid) => {
+        cfg[name] = cfg[name] || bid.params?.[name];
+      });
+    }
+  });
+  cfg.complete = FIELDS.every((field) => !field.required || cfg[field.name]);
+  if (cfg.complete) {
+    cfg.pbsHost = cfg.pbsHost.trim().replace('http://', 'https://');
+    if (cfg.pbsHost.indexOf('https://') < 0) {
+      cfg.pbsHost = `https://${cfg.pbsHost}`;
+    }
+  }
+  return cfg;
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+  gvlid: 1100,
+  supportedMediaTypes: [BANNER, VIDEO, NATIVE],
+
+  /** We need both params.placementId + a complete configuration (pbsHost + accountId) to continue **/
+  isBidRequestValid: (bid) => bid.params?.placementId && getBidderConfig([bid]).complete,
+
+  /** Trigger impression-pixel */
+  onBidWon: ({pbsWurl}) => pbsWurl && triggerPixel(pbsWurl),
+
+  /** Build BidRequest for PBS */
+  buildRequests(bidRequests, bidderRequest) {
+    const { bidder } = bidRequests[0];
+    const cfg = getBidderConfig(bidRequests);
+    const data = converter.toORTB({bidRequests, bidderRequest});
+
+    /** Set tmax, in general this will be timeout - pbsBufferMs */
+    const pbjsTimeout = bidderRequest.timeout || 1000;
+    data.tmax = Math.min(Math.max(pbjsTimeout - cfg.pbsBufferMs, cfg.pbsBufferMs), pbjsTimeout);
+
+    delete data.ext?.prebid?.aliases; // We don't need/want to send aliases to PBS
+    deepSetValue(data, 'ext.relevant', {
+      ...data.ext?.relevant,
+      adapter: true, // For internal analytics
+    });
+    deepSetValue(data, 'ext.prebid.storedrequest.id', cfg.accountId);
+    data.ext.prebid.passthrough = {
+      ...data.ext.prebid.passthrough,
+      relevant: { bidder }, // to find config for the right bidder-code in interpretResponse / getUserSyncs
+    };
+    return [{
+      method: 'POST',
+      url: `${cfg.pbsHost}/openrtb2/auction`,
+      data
+    }];
+  },
+
+  /** Read BidResponse from PBS and make necessary adjustments to not make it appear to come from unknown bidders */
+  interpretResponse(response, request) {
+    const resp = deepClone(response.body);
+    const { bidder } = request.data.ext.prebid.passthrough.relevant;
+    const editBid = (bid) => { // Modify bidder-codes in bid to the bidder code for this bidder
+      const { prebid = {} } = bid.ext || {};
+      const { meta, targeting } = prebid;
+      if (meta?.adaptercode) {
+        meta.adaptercode = bidder;
+      }
+      if (targeting?.hb_bidder) {
+        targeting.hb_bidder = bidder;
+      }
+      return bid;
+    };
```

Comment:
I'm not sure, this feels useful to discuss in committee. Why not just label all the bids that come back via this adapter as bidder-code relevant?

Reply:
Hi, I guess the other option would be that we would transform the response in the Prebid Server code instead. However, we really try hard to avoid own customisations of the Prebid Server source code.

So just to clarify a bit, what's happening here is that:

1. A normal Prebid Server instance is called (with stored request ids, so that the actual SSPs can be configured on the servers)
2. We get the response back
3. `interpretResponse` creates a copy of the server's `BidResponse` and modifies it so that it appears as the result of a single bidder/adapter - to not confuse analytic adapters / 3'rd party libraries, etc.
4. This involves changing `hb_bidder=relevantdigital` - in this copy of the server's response.
5. After this, the ORTB conversion library handles the rest by creating the response object using this modified server-response.

So we're _not_ changing the targeting in the bid's `adserverTargeting` object. Instead we're just doing some preprocessing of a copy of the server's response - before feeding it to ORTB covertion library's `fromORTB()` function.

Reply:
Following up on our offline discussion; I now understand what you're doing a bit better and I think it's OK. As you say, this logic is working on the response from your backend (and not on the "real" bid targeting as I - and maybe @jsnellbaker - initially thought).

Some of it can be made simpler; instead of modifying `seatbid.seat`  and `ext.prebid.meta.adaptercode` I would do:

```javascript
const converter = ortbConverter({
    // ... 
  overrides: {
    bidResponse: {
      bidderCode: false,
    }
  }
})
```

This would have the effect of turning off [this piece of logic](https://github.com/prebid/Prebid.js/blob/bff1f412f7e91e6eccc87b0779c8dce26288b0d2/libraries/pbsExtensions/processors/pbs.js#L46-L52), which is what's copying over `bidderCode` and `adapterCode` from the PBS response; without it, they default to whatever alias was used in the adUnit (which would also address that concern). To make it configurable:

```javascript
const converter = ortbConverter({
    // ... 
  overrides: {
    bidResponse: {
      bidderCode(orig, bidResponse, bid, context) {
        if (/* should bidder be the "true" bidder? */) {
          orig(bidResponse, bid, context); // run the "vanilla" logic (set bidder and adapter code as returned by PBS)
        }
      }
    }
  }
})
```

You could do something similar with [`extPrebidTargeting` or `extPrebidMeta`](https://github.com/prebid/Prebid.js/blob/bff1f412f7e91e6eccc87b0779c8dce26288b0d2/libraries/pbsExtensions/processors/pbs.js#L62-L76) but it may not be much simpler unless it makes sense to turn them off entirely.

Reply:
Added a `useSourceBidderCode` setting to let publishers chose to use the S2S bidder's code instead + used the nice trick proposed by @dgirardi.

The [docs PR](https://github.com/prebid/prebid.github.io/pull/4426) has now also been update.